lib/accesslists/walk.go

```diff
@@ -0,0 +1,305 @@
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	accesslistv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/accesslist/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/accesslist"
+	"github.com/gravitational/teleport/api/utils/clientutils"
+)
+
+var (
+	// Not using trace errors is intentional here as those errors are expected.
+	// They are used to control the graph traversal. Using trace.Errorf() or
+	// other trace errors would cause useless stacktrace captures each time an
+	// error is returned.
+	skipLeg = errors.New("skip this access leg")
+	skipAll = errors.New("skip everything and stop the walk")
+)
+
+// walkFunc is the type of the function called by walk on each
+// access list member and nested access list part of the access graph.
+// The error result returned by the function controls how accessPath continues.
+// If the function returns the special value SkipDir, Walk skips the current directory (path if info.IsDir() is true, otherwise path's parent directory). If the function returns the special value SkipAll, Walk skips all remaining files and directories. Otherwise, if the function returns a non-nil error, Walk stops entirely and returns that error.
+type walkFunc func(path accessPath) error
+
+// walkUntilUser returns a walkFunc that filters out every invalid
+// accessLeg. Invalid access legs are:
+// - expired legs
+// - legs granting access to a different user
+// - legs granting access to a list whose requirements are not met by the user
+func walkUntilUser(user types.User, now time.Time) (walkFunc, func() (accesslistv1.AccessListUserAssignmentType, error)) {
+	skipped := make([]skippedAccessPath, 0)
+	var result accesslistv1.AccessListUserAssignmentType
+
+	getResult := func() (accesslistv1.AccessListUserAssignmentType, error) {
+		if result == accesslistv1.AccessListUserAssignmentType_ACCESS_LIST_USER_ASSIGNMENT_TYPE_UNSPECIFIED {
+			// If we land here, no valid access paths were identified.
+			// To make troubleshooting easier, we optionally return a string
+			// explaining which access paths were filtered out.
+			return result, trace.AccessDenied("User is not member of the access list, directly or via nested list: %s", explainSkipped(skipped))
+		}
+		return result, nil
+	}
+
+	return func(path accessPath) error {
+		// First, skip the path if it  doesn't meet the requirements.
+		// This also prevents the walker from continuing to consider paths
+		// containing this leg, as they would be invalid as well.
+
+		// Because the walkFunc was already called on every leg of the path,
+		// we only have to consider the last one.
+		leg := path[len(path)-1]
+		if leg.member != nil {
+			// If the membership is for a user but not the one we are looking for, we skip it.
+			if leg.member.Spec.MembershipKind == accesslist.MembershipKindUser && leg.member.Spec.Name != user.GetName() {
+				return skipLeg
```

Comment:
That reads a bit strange. Shouldn't that return nil? Is there anything to skip for a user member? See also https://github.com/gravitational/teleport/pull/60034/files#r2558046225